### PR TITLE
Shoshi/profile page

### PR DESCRIPTION
**Pull request by Shoshana-Yudelevitz**
עשיתי את המשימה add pagination for profiles table, display "מספר שמתקבל" profiles for a page
בעמוד   TableComponent שמתאימה לכל מספר שיתקבל, ומוגדרת כברירת מחדל.
ועדכנתי את העמוד ProfilePageComponent שישלח את המספר 5 כמו שביקשו במשימה.

### CODE REVIEW

**Review by avishag-gvili — APPROVED** [review]

File: client-side/src/components/profileComponents/addProfileComponent.jsx [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+import { Button, Dialog, DialogActions, DialogContent, DialogContentText, DialogTitle, Tooltip } from '@mui/material';
+import { useSnackbar } from 'notistack';
+import GenericButton from '../../stories/Button/GenericButton.jsx';
+import GenericInput from '../../stories/GenericInput/genericInput.jsx';
+import { addProfile } from '../../redux/profile/profile.slice.js';
+import { createProfile } from '../../services/profileService.js';
+import RadioButton from '../../stories/RadioButton/radio-Button.jsx';
+import { SELECT_OPTIONS, INPUT_LABELS, DIALOG_TITLES, TOAST_MESSAGES, VALIDATE_MESSAGES, BUTTON_LABELS} from '../../constants/profileConstants.js';
+import '../../styles/profilePageStyle.scss';
+import ToastMessage from '../../stories/Toast/ToastMessage.jsx';
```
Comment (e171271) [inline client-side/src/components/profileComponents/addProfileComponent.jsx:13]: please re-order the imports

File: client-side/src/components/profileComponents/addProfileComponent.jsx [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+export default function AddProfile({ userId }) {
+  const dispatch = useDispatch();
+  const navigate = useNavigate();
+  const { enqueueSnackbar } = useSnackbar();
+  const [open, setOpen] = React.useState(false);
+  const [data, setData] = React.useState(getInitialData());
+  const [errorText, setErrorText] = React.useState('');
```
Comment (e171271) [inline client-side/src/components/profileComponents/addProfileComponent.jsx:21]: please import useState in the react import line and use it here

File: client-side/src/components/profileComponents/addProfileComponent.jsx [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+      timeStart: '00:00',
+      timeEnd: '00:00',
+      status: '',
+      url: '',
+      urlTimeLimit: 0,
+      urlStatus: ''
+    };
+  }
+
+  const handleClose = useCallback(() => {
```
Comment (e171271) [inline client-side/src/components/profileComponents/addProfileComponent.jsx:35]: why do you need useCallback? - relevant for all functions

File: client-side/src/components/profileComponents/addProfileComponent.jsx [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    } catch (error) {
+      console.error(TOAST_MESSAGES.PROFILE_CREATE_ERROR, error);
+      enqueueSnackbar(<ToastMessage message={TOAST_MESSAGES.PROFILE_CREATE_ERROR} type="error" />);
+    }
+  }, [data, dispatch, navigate, handleClose, enqueueSnackbar]);
+
+ 
+
+  return (
+    <React.Fragment>
```
Comment (e171271) [inline client-side/src/components/profileComponents/addProfileComponent.jsx:95]: why do you need fragment?

File: client-side/src/components/profileComponents/addProfileComponent.jsx [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+          />
+          <DialogContentText className='dialog-content-text'>{DIALOG_TITLES.PROFILE_TIME} </DialogContentText>
+          <div className='div-time'>
+            <GenericInput
+              label={INPUT_LABELS.TIME_START}
+              name="timeStart"
+              type="time"
+              value={data.timeStart}
+              onChange={handleChange}
+              width='100%'
```
Comment (e171271) [inline client-side/src/components/profileComponents/addProfileComponent.jsx:131]: why do you have styles here? it should be in scss file

File: client-side/src/components/profileComponents/addProfileComponent.jsx [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+              onChange={handleChange}
+            />
+           
+          </div> 
+        </DialogContent>
+        <DialogActions>
+          <Button color="error" onClick={handleClose}>
+            {BUTTON_LABELS.CANCEL}
+          </Button>
+          {(!data.name || data.name.length < 2 || data.name.length > 50 || !data.status) ? (
```
Comment (e171271) [inline client-side/src/components/profileComponents/addProfileComponent.jsx:157]: please create a variable for this condition and use it here

File: client-side/src/components/profileComponents/addProfileComponent.jsx [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            <Tooltip title={TOAST_MESSAGES.FORM_NOT_FILLED}>
+              <span>
+                <Button color="success" type="submit" disabled={!data.name || data.name.length < 2 || data.name.length > 50 || !data.status}>
```
Comment (e171271) [inline client-side/src/components/profileComponents/addProfileComponent.jsx:160]: please create a variable for this condition and use it here

File: client-side/src/components/profileComponents/profileActivationComponent.jsx [hunk 8]
```diff
@@ -0,0 +1,39 @@
+import React from 'react';
+import { CountdownCircleTimer } from 'react-countdown-circle-timer';
+import '../../styles/profileActivationTimer.scss';
+
+const ProfileActivationTimer = ({ profileActivationTime }) => {
```
Comment (e171271) [inline client-side/src/components/profileComponents/profileActivationComponent.jsx:5]: please add default value

File: client-side/src/components/profileComponents/profileActivationComponent.jsx [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    const hours = Math.floor(remainingTime / 3600);
+    const minutes = Math.floor((remainingTime % 3600) / 60);
+    const seconds = remainingTime % 60;
+    const formattedMinutes = minutes < 10 ? `0${minutes}` : `${minutes}`;
+    const formattedSeconds = seconds < 10 ? `0${seconds}` : `${seconds}`;
+
+    return (
+      <div className="timer-container">
+        <div className="timer-time">{`${hours}:${formattedMinutes}:${formattedSeconds}`}</div>
+        <div className="timer-label">Time Left</div>
```
Comment (e171271) [inline client-side/src/components/profileComponents/profileActivationComponent.jsx:18]: please move text to constants file

File: client-side/src/components/profileComponents/profileActivationComponent.jsx [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        size={100}
+        trailColor="#d6d6d6"
+      >
+        {renderTime}
+      </CountdownCircleTimer>
+    </div>
+  );
+};
+
+export default ProfileActivationTimer;
```
Comment (e171271) [inline client-side/src/components/profileComponents/profileActivationComponent.jsx:39]: please add prop-ytpe

File: client-side/src/components/profileComponents/profileActivationComponent.jsx [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+  return (
+    <div className="timer-wrapper">
+      <CountdownCircleTimer
+        isPlaying
+        duration={durationSeconds}
+        colors={[['rgb(45, 158, 126)']]}
+        strokeWidth={4}
+        size={100}
+        trailColor="#d6d6d6"
```
Comment (e171271) [inline client-side/src/components/profileComponents/profileActivationComponent.jsx:31]: move all styles to scss file

File: client-side/src/components/profileComponents/profilePageComponent.jsx [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+import { setProfiles, updateProfile } from '../../redux/profile/profile.slice.js';
+import { selectProfile } from '../../redux/profile/profile.selector.js';
+import { Delete as DeleteIcon, Edit as EditIcon, Save as SaveIcon, Cancel as CancelIcon } from '@mui/icons-material';
+import Loader from '../../stories/loader/loader.jsx';
+import AddProfileComponent from './addProfileComponent.jsx';
+import UpdateProfileComponent from './updateProfileCpmponent.jsx';
+import ProfileActivationTimer from './profileActivationComponent.jsx';
+import { extractWebsiteName, isValidURL, isWebsiteInProfile, getStatusOptions } from '../../utils/profileUtil.js';
+import { TOAST_MESSAGES } from '../../constants/profileConstants.js';
+import '../../styles/profilePageStyle.scss';
```
Comment (e171271) [inline client-side/src/components/profileComponents/profilePageComponent.jsx:19]: please re-order the imports

File: client-side/src/components/profileComponents/profilePageComponent.jsx [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+
+const ProfilePageComponent = ({ userId }) => {
```
Comment (e171271) [inline client-side/src/components/profileComponents/profilePageComponent.jsx:21]: please add default value

File: client-side/src/components/profileComponents/profilePageComponent.jsx [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+
+  const handleDelete = async (id) => {
+    const updatedWebsites = selectedProfile.listWebsites.filter(website => website.websiteId._id !== id);
+    const profileToUpdate = {
+      ...selectedProfile,
+      listWebsites: updatedWebsites
+    };
+    try {
+      await deleteWebsite(id);
+      await updateProfileApi(selectedProfile._id, profileToUpdate);
```
Comment (e171271) [inline client-side/src/components/profileComponents/profilePageComponent.jsx:82]: maybe you can join the both server calls to one,for decrease time?

File: client-side/src/components/profileComponents/profilePageComponent.jsx [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+      dispatch(updateProfile(profileToUpdate));
```
Comment (e171271) [inline client-side/src/components/profileComponents/profilePageComponent.jsx:83]: you should dispatch only if the server call succeed

File: client-side/src/components/profileComponents/profilePageComponent.jsx [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+    }
+    if ((editRowId === 'new') && (editedRows.url === '' || editedRows.status === '')) {
+      enqueueSnackbar(<ToastMessage message={TOAST_MESSAGES.WEBSITE_SAVE_ERROR} type="error" />);
+      return;
+    }
+    if (!isValidURL(editedRows.url)) {
+      enqueueSnackbar(<ToastMessage message={TOAST_MESSAGES.INVALID_URL} type="error" />);
+      return;
+    }
+    if (editedRows.id==='new'&&(isWebsiteInProfile(editedRows.url, selectedProfile))) {
```
Comment (e171271) [inline client-side/src/components/profileComponents/profilePageComponent.jsx:118]: please change all the if else to switch case

File: client-side/src/router/router.jsx [hunk 17]
```diff
@@ -17,8 +17,9 @@ export  const router = createBrowserRouter([
             },
             {
                 path: '/profiles',
-                element:<ProfileList/>
-            }
+                element:<ProfilePageComponent  userId = {'6698da056e5c07ebd3c11ec1'}/>
+
```
Comment (e171271) [inline client-side/src/router/router.jsx:21]: don't use hard-code values

File: client-side/src/stories/stylies.scss [hunk 18]
```diff
@@ -0,0 +1,5 @@
+$background-class: #ffffff;
+$secondary-color-hard:  rgb(200, 245, 232);
+$primary-color: black;
+$secondary-color: rgb(103, 252, 210);
+@import url('https://fonts.googleapis.com/css2?family=Nanum+Gothic&display=swap');
```
Comment (e171271) [inline client-side/src/stories/stylies.scss:5]: please move all variables to variables file, this stiles file already deleted from main branch

File: client-side/src/stories/table/TableComponent.jsx [hunk 19]
```diff
@@ -1,38 +1,166 @@
...
 import React from 'react';
 import { DataGrid } from '@mui/x-data-grid';
-import PropTypes from 'prop-types'
+import PropTypes from 'prop-types';
+import IconButton from '@mui/material/IconButton';
+import TextField from '@mui/material/TextField';
+import Select from '@mui/material/Select';
+import MenuItem from '@mui/material/MenuItem';
+import Tooltip from '@mui/material/Tooltip';
+import GenericButton from '../Button/GenericButton.jsx';
```
Comment (e171271) [inline client-side/src/stories/table/TableComponent.jsx:10]: please join all mui imports to one

File: client-side/src/stories/table/TableComponent.jsx [hunk 20]
```diff
@@ -1,38 +1,166 @@
...
+  dataObject,
+  widthOfTable,
+  widthOfColums,
+  actions,
+  editRowId,
+  handleFieldChange,
+  statusOptions,
+  addButton,
+  handleAddRow,
+  pageSize
```
Comment (e171271) [inline client-side/src/stories/table/TableComponent.jsx:23]: please add default values to each one that not marked as required

File: client-side/src/stories/table/TableComponent.jsx [hunk 21]
```diff
@@ -1,38 +1,166 @@
...
     field: header,
     headerName: header,
     width: widthOfColums[i],
-    align:'center',
-    headerAlign: 'center'
+    align: 'center',
+    headerAlign: 'center',
+    renderCell: (params) => {
+      if (editRowId && editRowId === params.row.id && header !== 'Actions') {
+        if (header === 'status') {
```
Comment (e171271) [inline client-side/src/stories/table/TableComponent.jsx:33]: please change to switch case

File: client-side/src/stories/table/TableComponent.jsx [hunk 22]
```diff
@@ -1,38 +1,166 @@
...
+                <MenuItem key={option.value} value={option.value}>
+                  {option.text}
+                </MenuItem>
+              ))}
+            </Select>
+          );
+        }
+        if (header === 'name') {
+          return (
+            <Tooltip title='This field is optional'>
```
Comment (e171271) [inline client-side/src/stories/table/TableComponent.jsx:51]: please move text to constants file - do it in all your code

File: client-side/src/stories/table/TableComponent.jsx [hunk 23]
```diff
@@ -1,38 +1,166 @@
...
+        );
+      }
+      return params.value;
+    },
   }));
 
+  if (!columns.some(column => column.field === 'Actions')) {
+    columns.push({
+      field: 'Actions',
+      headerName: 'Actions',
```
Comment (e171271) [inline client-side/src/stories/table/TableComponent.jsx:90]: what is the difference between this and what you have in line 70 - 83, are they the same?

File: client-side/src/stories/table/TableComponent.jsx [hunk 24]
```diff
@@ -1,38 +1,166 @@
...
+            </Tooltip>
+          ))}
+        </div>
+      ),
+    });
+  }
+
   return (
-    <div className="table" style={{ width: widthOfTable ,marginTop:"8%"}}>
+    <div className="table" style={{ width: widthOfTable, marginTop: "8%" }}>
```
Comment (e171271) [inline client-side/src/stories/table/TableComponent.jsx:109]: move all styles to scss file - do it in all your code

File: client-side/src/stories/table/TableComponent.jsx [hunk 25]
```diff
@@ -1,38 +1,166 @@
...
+    value: PropTypes.string.isRequired,
+    text: PropTypes.string.isRequired,
+  })).isRequired,
+  addButton: PropTypes.bool,
+  handleAddRow: PropTypes.func,
+  pageSize: PropTypes.number, 
 };
 
-export default TableComponent;
+TableComponent.defaultProps = {
```
Comment (e171271) [inline client-side/src/stories/table/TableComponent.jsx:156]: please remove defaultProps and add default values instead

File: client-side/src/styles/profileActivationTimer.scss [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+.timer-label {
+  font-size: 60%;
+  margin-top: 4%;
+  color: $secondary-color;
+}
+
+.countdown-circle-timer {
+  stroke-width: 0.25%;
+  size: 6.25%;
+//   trail-color: $standardInfoColor; // Use standard info color for trail
```
Comment (e171271) [inline client-side/src/styles/profileActivationTimer.scss:31]: please remove comments

File: client-side/src/styles/profileActivationTimer.scss [hunk 27]
```diff
@@ -0,0 +1,33 @@
+@import url('https://fonts.googleapis.com/css2?family=Nanum+Gothic&display=swap');
+@import '../stories/variables.scss';
+
+
+.timer-wrapper {
```
Comment (e171271) [inline client-side/src/styles/profileActivationTimer.scss:5]: this file is looks like css file, scss file should have hierarchy.

File: client-side/src/styles/profilePageStyle.scss [hunk 28]
```diff
@@ -0,0 +1,120 @@
...
+    }
+    &:focus {
+      outline: none;
+    }
+  }
+}
+.helper-text-error {
+  color: $standardErrorColor;
+}
+
```
Comment (e171271) [inline client-side/src/styles/profilePageStyle.scss:111]: this file is looks like css file, scss file should have hierarchy.

File: client-side/src/styles/profilePageStyle.scss [hunk 29]
```diff
@@ -0,0 +1,120 @@
...
+.profile-select {
+  width: 10vw;
+}
+.timer {
+  margin-left:3vh;
+}
+.green_title{
+  color:  rgb(103, 252, 210) ;
```
Comment (e171271) [inline client-side/src/styles/profilePageStyle.scss:119]: please move all colors and fonts to variables

File: server-side/controllers/profile.controller.js [hunk 30]
```diff
@@ -1,77 +1,81 @@
-import mongoose  from 'mongoose';
-import Profiles from '../models/profile.model.js';
+import Profile from '../models/profile.model.js';
 
-
-export const getAllProfiles = async (req, res, next) => {
+export const getAllProfiles = async (req, res) => {
     try {
-        const profiles = await Profiles.find().populate('limitedWebsites.websiteId blockedSites').select('-__v');
+        const profiles = await Profile.find();
```
Comment (e171271) [inline server-side/controllers/profile.controller.js:5]: you shouldn't fetch from the collection from the collection, you should do it from service file - do it in all your code

**Review by e171271 — CHANGES_REQUESTED** [review]
I review for step 1, after you will fix all I'll move over the code again

**Pull request closed without merging**